cli/ts/commands/joinPoll.ts

```diff
@@ -0,0 +1,257 @@
+  newVoiceCreditBalance,
+  stateFile,
+  pollId,
+  signer,
+  startBlock,
+  endBlock,
+  blocksPerBatch,
+  transactionHash,
+  pollJoiningZkey,
+  useWasm,
+  rapidsnark,
+  pollWitgen,
+  pollWasm,
+  quiet = true,
+}: IJoinPollArgs): Promise<number> => {
+  banner(quiet);
+  const userSideOnly = true;
+
+  if (!(await contractExists(signer.provider!, maciAddress))) {
+    logError("MACI contract does not exist");
+  }
+
+  if (!PrivKey.isValidSerializedPrivKey(privateKey)) {
+    logError("Invalid MACI private key");
+  }
+
+  const userMaciPrivKey = PrivKey.deserialize(privateKey);
+  const nullifier = poseidon([BigInt(userMaciPrivKey.asCircuitInputs())]);
+
+  // Create poll public key from poll private key
+  const pollPrivKeyDeserialized = PrivKey.deserialize(pollPrivKey);
+  const pollKeyPair = new Keypair(pollPrivKeyDeserialized);
+  const pollPubKey = pollKeyPair.pubKey;
+
+  // check < 1 cause index zero is a blank state leaf
+  if (stateIndex < 1) {
+    logError("Invalid state index");
+  }
+
+  if (pollId < 0) {
+    logError("Invalid poll id");
+  }
+
+  const maciContract = MACIFactory.connect(maciAddress, signer);
+  const pollAddress = await maciContract.getPoll(pollId);
+
+  if (!(await contractExists(signer.provider!, pollAddress))) {
+    logError("Poll contract does not exist");
+  }
+
+  const pollContract = PollFactory.connect(pollAddress, signer);
+
+  let maciState: MaciState | undefined;
+  if (stateFile) {
+    const content = JSON.parse(fs.readFileSync(stateFile).toString()) as unknown as IJsonMaciState;
+
+    try {
+      maciState = MaciState.fromJSON(content);
+    } catch (error) {
+      logError((error as Error).message);
+    }
+  } else {
+    // build an off-chain representation of the MACI contract using data in the contract storage
+    const [defaultStartBlockSignup, defaultStartBlockPoll, stateRoot, numSignups] = await Promise.all([
+      maciContract.queryFilter(maciContract.filters.SignUp(), startBlock).then((events) => events[0]?.blockNumber ?? 0),
+      maciContract
+        .queryFilter(maciContract.filters.DeployPoll(), startBlock)
+        .then((events) => events[0]?.blockNumber ?? 0),
+      maciContract.getStateTreeRoot(),
+      maciContract.numSignUps(),
+    ]);
+    const defaultStartBlock = Math.min(defaultStartBlockPoll, defaultStartBlockSignup);
+    let fromBlock = startBlock ? Number(startBlock) : defaultStartBlock;
+
+    const defaultEndBlock = await Promise.all([
+      pollContract
+        .queryFilter(pollContract.filters.MergeMaciState(stateRoot, numSignups), fromBlock)
+        .then((events) => events[events.length - 1]?.blockNumber),
+    ]).then((blocks) => Math.max(...blocks));
+
+    if (transactionHash) {
+      const tx = await signer.provider!.getTransaction(transactionHash);
+      fromBlock = tx?.blockNumber ?? defaultStartBlock;
+    }
+
+    logYellow(quiet, info(`starting to fetch logs from block ${fromBlock}`));
+    // TODO: create genPollStateTree ?
```

Comment:
https://github.com/0x3327/maci/commit/942596989352aceca2394adaf16b04d442140ef0

- removed `genMaciStateFromContract`
- removed `userSideOnly` in  `genMaciStateFromContract` for verifying coordinator keys
- updated `genSignUpTree` to return sign up tree and state leaves
- copied `joiningCircuitInputs` to joinPoll and use it instead of `poll.joiningCircuitInputs`